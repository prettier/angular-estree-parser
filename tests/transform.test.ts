import type * as ng from '@angular/compiler';
import type * as b from '@babel/types';
import {
  parseAction,
  parseBinding,
  parseInterpolation,
  parseSimpleBinding,
} from '../src/index.js';
import type { NGNode } from '../src/types.js';
import {
  getNgType,
  parseNgAction,
  parseNgBinding,
  parseNgInterpolation,
  parseNgSimpleBinding,
} from '../src/utils.js';
import {
  massageAst,
  parseBabel,
  parseBabelExpression,
  snapshotAst,
} from './helpers.js';

describe.each`
  beforeType            | afterType                     | input                       | action   | binding  | simple   | interpolation
  ${'Binary'}           | ${'BinaryExpression'}         | ${' 0 - 1 '}                | ${true}  | ${true}  | ${true}  | ${true}
  ${'Binary'}           | ${'LogicalExpression'}        | ${' a && b '}               | ${true}  | ${true}  | ${true}  | ${true}
  ${'Unary'}            | ${'UnaryExpression'}          | ${' - 1 '}                  | ${true}  | ${true}  | ${true}  | ${true}
  ${'Unary'}            | ${'UnaryExpression'}          | ${' + 1 '}                  | ${true}  | ${true}  | ${true}  | ${true}
  ${'BindingPipe'}      | ${'NGPipeExpression'}         | ${' a | b '}                | ${false} | ${true}  | ${false} | ${true}
  ${'BindingPipe'}      | ${'NGPipeExpression'}         | ${' a | b : c '}            | ${false} | ${true}  | ${false} | ${true}
  ${'Chain'}            | ${'NGChainedExpression'}      | ${' a ; b '}                | ${true}  | ${false} | ${false} | ${false}
  ${'Conditional'}      | ${'ConditionalExpression'}    | ${' a ? 1 : 2 '}            | ${true}  | ${true}  | ${true}  | ${true}
  ${'EmptyExpr'}        | ${'NGEmptyExpression'}        | ${''}                       | ${true}  | ${true}  | ${true}  | ${false}
  ${'Call'}             | ${'CallExpression'}           | ${' ( a . b ) ( 1 , 2 ) '}  | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' ( a . b )?.( 1 , 2 ) '} | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' ( a ) ( 1 , 2 ) '}      | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' ( a )?.( 1 , 2 ) '}     | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' a ( 1 ) ( 2 ) '}        | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ( 1 )?.( 2 ) '}       | ${true}  | ${true}  | ${true}  | ${true}
  ${'KeyedRead'}        | ${'MemberExpression'}         | ${' a [ b ] '}              | ${true}  | ${true}  | ${true}  | ${true}
  ${'KeyedRead'}        | ${'OptionalMemberExpression'} | ${' a ?. b [ c ] '}         | ${true}  | ${true}  | ${true}  | ${true}
  ${'KeyedRead'}        | ${'OptionalMemberExpression'} | ${' a ?. b () [ c ] '}      | ${true}  | ${true}  | ${true}  | ${true}
  ${'KeyedWrite'}       | ${'AssignmentExpression'}     | ${' a [ b ] = 1 '}          | ${true}  | ${true}  | ${true}  | ${true}
  ${'ImplicitReceiver'} | ${'ThisExpression'}           | ${' this '}                 | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralArray'}     | ${'ArrayExpression'}          | ${' [ 1 ] '}                | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralMap'}       | ${'ObjectExpression'}         | ${' { "a" : 1 } '}          | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralMap'}       | ${'ObjectExpression'}         | ${' ( { a : 1 } ) '}        | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' f ( { a : 1 } ) '}      | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralMap'}       | ${'ObjectExpression'}         | ${' ( {a, b: 2} ) '}        | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' f ( {a, b: 2} ) '}      | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralMap'}       | ${'ObjectExpression'}         | ${' ( {a, b} ) '}           | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralMap'}       | ${'ObjectExpression'}         | ${' ( { a, b} ) '}          | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'BooleanLiteral'}           | ${' true '}                 | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'Identifier'}               | ${' undefined '}            | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'NullLiteral'}              | ${' null '}                 | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'NumericLiteral'}           | ${' ( 1 ) '}                | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'NumericLiteral'}           | ${' 1 '}                    | ${true}  | ${true}  | ${true}  | ${true}
  ${'LiteralPrimitive'} | ${'StringLiteral'}            | ${' "hello" '}              | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' a ( this ) '}           | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?.( this ) '}         | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' a ( b) '}               | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?.( b) '}             | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' a . b ( 1 , 2 ) '}      | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a . b ?.( 1 , 2 ) '}    | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'CallExpression'}           | ${' a ( 1 , 2 ) '}          | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?. ( 1 , 2 ) '}       | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'OptionalCallExpression'}   | ${' a ?. b . c ( ) '}       | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?. b . c ?. ( ) '}    | ${true}  | ${true}  | ${true}  | ${true}
  ${'Call'}             | ${'OptionalCallExpression'}   | ${' a ?. b ( ) . c ( ) '}   | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?. b ( ) . c ?.( ) '} | ${true}  | ${true}  | ${true}  | ${true}
  ${'NonNullAssert'}    | ${'TSNonNullExpression'}      | ${' x ! '}                  | ${true}  | ${true}  | ${true}  | ${true}
  ${'PrefixNot'}        | ${'UnaryExpression'}          | ${' ! x '}                  | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'Identifier'}               | ${' ( ( a ) ) '}            | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'Identifier'}               | ${' a '}                    | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'Identifier'}               | ${' a // hello '}           | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'MemberExpression'}         | ${' a . b '}                | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'MemberExpression'}         | ${' this . a '}             | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'OptionalMemberExpression'} | ${' a ?. b . c '}           | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyRead'}     | ${'OptionalMemberExpression'} | ${' a ?. b ( ) . c '}       | ${true}  | ${true}  | ${true}  | ${true}
  ${'PropertyWrite'}    | ${'AssignmentExpression'}     | ${' a . b = 1 '}            | ${true}  | ${false} | ${false} | ${false}
  ${'PropertyWrite'}    | ${'AssignmentExpression'}     | ${' a = 1 '}                | ${true}  | ${false} | ${false} | ${false}
  ${'Call'}             | ${'OptionalCallExpression'}   | ${' a ?. b ( ) '}           | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafeCall'}         | ${'OptionalCallExpression'}   | ${' a ?. b ?. ( ) '}        | ${true}  | ${true}  | ${true}  | ${true}
  ${'SafePropertyRead'} | ${'OptionalMemberExpression'} | ${' a ?. b '}               | ${true}  | ${true}  | ${true}  | ${true}
`('$input ($beforeType -> $afterType)', (fields) => {
  const { beforeType, afterType, input } = fields;

  let beforeNode: ng.AST | null = null;
  let afterNode: NGNode | null = null;

  const testSection = (
    section: Extract<keyof typeof fields, string>,
    parseBefore: (input: string) => { ast: ng.AST },
    parseAfter: (input: string) => NGNode,
  ) => {
    if (fields[section]) {
      test(`allowed in ${section}`, () => {
        expect(() => (beforeNode = parseBefore(input).ast)).not.toThrowError();
        expect(() => (afterNode = parseAfter(input))).not.toThrowError();
      });
    } else {
      test(`disallowed in ${section}`, () => {
        expect(() => parseBefore(input)).toThrowError();
        expect(() => parseAfter(input)).toThrowError();
      });
    }
  };

  testSection('action', parseNgAction, parseAction);
  testSection('binding', parseNgBinding, parseBinding);
  testSection('simple', parseNgSimpleBinding, parseSimpleBinding);
  testSection('interpolation', parseNgInterpolation, parseInterpolation);

  test('ast', () => {
    expect(beforeNode).not.toEqual(null);
    expect(afterNode).not.toEqual(null);

    expect(getNgType(beforeNode!)).toEqual(beforeType);
    expect(afterNode!.type).toEqual(afterType);

    if (afterNode!.type.startsWith('NG')) {
      expect(snapshotAst(afterNode, input)).toMatchSnapshot();
    } else {
      try {
        expect(afterNode).toEqual(massageAst(parseBabelExpression(input)));
      } catch {
        const { comments, program } = parseBabel(input);
        const statement = program.body[0] as b.ExpressionStatement;
        expect(statement.type).toEqual('ExpressionStatement');
        expect(massageAst(afterNode)).toEqual(
          massageAst({ ...statement.expression, comments }),
        );
      }
    }
  });
});
