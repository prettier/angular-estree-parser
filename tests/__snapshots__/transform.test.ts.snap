// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` (EmptyExpr -> NGEmptyExpression) ast 1`] = `
NGEmptyExpression { comments: [] }
> 1 |
    | ^
`;

exports[` a ; b  (Chain -> NGChainedExpression) ast 1`] = `
NGChainedExpression { expressions: ["Identifier", "Identifier"], comments: [] }
> 1 |  a ; b 
    |  ^^^^^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a ; b 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "b" }
> 1 |  a ; b 
    |      ^
`;

exports[` a | b  (BindingPipe -> NGPipeExpression) ast 1`] = `
NGPipeExpression { left: "Identifier", right: "Identifier", arguments: [], comments: [] }
> 1 |  a | b 
    |  ^^^^^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a | b 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "b" }
> 1 |  a | b 
    |      ^
`;

exports[` a | b : c  (BindingPipe -> NGPipeExpression) ast 1`] = `
NGPipeExpression {
  left: "Identifier",
  right: "Identifier",
  arguments: ["Identifier"],
  comments: [],
}
> 1 |  a | b : c 
    |  ^^^^^^^^^
--------------------------------------------------------------------------------
Identifier { name: "a" }
> 1 |  a | b : c 
    |  ^
--------------------------------------------------------------------------------
Identifier { name: "b" }
> 1 |  a | b : c 
    |      ^
--------------------------------------------------------------------------------
Identifier { name: "c" }
> 1 |  a | b : c 
    |          ^
`;
